### ISSUE TITLE
Memory issues within Docker containers.

### ISSUE THREAD

**Issue by GuillaumeRochette** [jnwatson/py-lmdb#225]
### Affected Operating Systems

* Personal Computer: Ubuntu 18.04
* Server: Docker: Ubuntu 16.04 (but can be moved to 18.04)

### Affected py-lmdb Version

0.98

### py-lmdb Installation Method

Within a conda environment. 
```
pip install lmdb
```

### Using bundled or distribution-provided LMDB library?

Bundled. (If that means the binaries that come with the pip install)

### Distribution name and LMDB library version

(0, 9, 22)

### Machine "free -h" output

For the machine I'm currently using:
```
              total        used        free      shared  buff/cache   available
Mem:           7,7G        2,1G        1,2G        173M        4,4G        5,1G
Swap:          7,6G        133M        7,5G
```
For one of the remote servers I have access:
```
              total        used        free      shared  buff/cache   available
Mem:           187G         43G         59G        577M         84G        140G
Swap:          7.6G        7.6G          0B
```
### Other important machine info

Regarding the servers, they're using a StorNext file system, but it is supposed to blend seamlessly with any local file system.
Also, the way one interacts with these servers is done via HTCondor, which starts Docker container, with limits on RAM, CPUs, GPUs, etc.

### Describe Your Problem

I'm working in research in Computer Vision and I'm using LMDB to access data quickly and efficiently to train models.

For example, I have multiple sequences, which contains multiple videos. From each of these videos, frames are extracted and stored in an LMDB. At the moment, I have one LMDB per video. It might not be optimal, but it is the most flexible option.

To train such model, I use PyTorch, which uses some multiprocessing, so that you can have multiple workers fetching the data to the GPUs.

Here is my wrapper for the database to be accessed in a read-only:
```py
import lmdb
import pickle

class Database(object):
    def __init__(self, path: str):
        self.db = lmdb.open(path, readonly=True, readahead=False, lock=False)

        with self.db.begin() as txn:
            indexes = pickle.loads(txn.get(key=pickle.dumps("indexes")))
        self.indexes = set(indexes)

    def __iter__(self):
        return iter(self.indexes)

    def __len__(self):
        return len(self.indexes)

    def __getitem__(self, index):
        if index not in self.indexes:
            raise KeyError(index)

        key = pickle.dumps(index)
        with self.db.begin() as txn:
            value = txn.get(key=key)
        return value

    def __del__(self):
        self.db.close()
```
The structure of a dataset is the following:
```py
import os.path as osp

from torch.utils.data import Dataset

from database import Database

class MyDataset(Dataset):
    def __init__(
        self,
        sequence: str,
        transform=None,
    ):
        paths = {
            0: osp.join(sequence, "0.lmdb"),
            1: osp.join(sequence, "1.lmdb"),
            ...
        }
        self.views = [view for view in images]
        self.images = {view: Database(paths[view]) for view in self.views}
        self.items = self._items()
        self.transform = transform

    def _indexes(self):
        indexes = [self.images[view].indexes for view in self.views]
        indexes = sorted(indexes[0].intersection(*(indexes[1:])))
        return indexes

    def _pairs(self):
        pairs = [[view_A, view_B] for view_A in self.views for view_B in self.views]
        return pairs

    def _items(self):
        items = [[index, pair] for index in self._indexes() for pair in self._pairs()]
        return items

    def __len__(self):
        return len(self.items)

    def __getitem__(self, item):
        index, (view_A, view_B) = self.items[item]
        frame = {
            "A": self.images[view_A][index]
            "B": self.images[view_B][index],
        }
        if self.transform:
            frame = self.transform(frame)
        return frame

```

### Interrogations

On the machine I'm currently using, retrieving data using the object MyDataset works fine both in single-threaded mode and multi-threaded, e.g. within a DataLoader, which handles the multiprocessing part with the workers.

![Memory](https://user-images.githubusercontent.com/11836548/71422185-2d8b6180-2680-11ea-8987-6769281dc189.png)
Here's the graph of the memory usage reported by **mprof** (memory_profiler). First of all, I see that the memory usage plateaus, but I'm not really sure why, I expected it to stay very low, unless it has to do with some caching ?
Second, **htop** and **free** are reporting different informations. During the execution of the script for the shared memory, regarding the workers processes **htop** reports a usage which increases over time, starting at 100MB and finishing around 400MB, while **free** shows an overall usage starting at 160MB and oscillating around 170MB. Again, I cannot really explain why ...

Regarding the servers, things are a slightly different. We are aiming for performance and not testing/debugging, therefore the objective is to have a 100% of the GPU cores being used constantly. All the pre-processing, e.g. image resizing, cropping, etc., is done on CPUs within the **transform**, and the idea is to increase the amount of workers until we reach that 100% GPU usage.

On the servers, one request some resources (CPU, RAM, GPU), and your program is executed, but if you happens to require more RAM than reserved, it raises an error and the job is terminated.
I also recall hearing that for these servers that the maximum amount of shared memory given to the docker container is equal to the amount of RAM one requests. 
I don't know if that is linked to the problem or if something else, with the cache is happening, but for some reason, when I have a low number of workers, jobs are running smoothly, but their shared memory value reported by **htop** is growing huge to around 14-15GB, when the RAM requested is set to 16GB.

Basically, I would like know if you have an idea on what is going wrong in here ? And where I should investigate before so that I can provide a solution to the people in charge of maintaining these servers ?

Thanks in advance,

Guillaume Rochette

**Comment by jnwatson** [comment]
Most memory profiling tools don't appropriately account for memory-mapped files, which is what lmdb (and ElasticSearch, and LevelDB and others) uses. Memory-mapped files will appear in virtual, resident, and shared columns in ps for example. Memory mapped files are not directly requests for RAM, and don't reduce the available RAM on the system, but the system for accounting for it is frankly pretty bad in LInux.

Essentially, you can't trust any of your measurement tools for RAM usage. You can use a tool to parse /proc/<pid>/smaps to get a better handle of where your RAM usage is going.

**Comment by GuillaumeRochette** [comment]
Thank you for that information. 
Do you reckon that cgroups, because that's what Docker uses to report memory usage, does not measure accurately as well?

Additionally, I have been using ```pmap -x <pid>``` on each individual thread to see their memory consumption.
Is there a way to compute the grand total of the physical memory used by a process which spawns multiple threads? And if so, is it extendable to multiple processes which themselves spawn threads?

Thanks a lot,

Guillaume Rochette

**Comment by jnwatson** [comment]
That's an excellent question. There's surprisingly not a whole lot of information around the interaction between cgroups and memory mapped files.

I did find this page from TerarkDB, a fork of RocksDB, another database that uses large memory maps.

https://github.com/Terark/terarkdb/wiki/Notes-about-mmap-and-cgroup

> This means that, if we access a file by read(read existing file) or write(write new file) system call, and mmap this file later, we can bypass the cgroup memory limit

Still in the normal case where there's only one process using the lmdb file, it looks like it will be accounted for correctly.

On your second question, I am somewhat confused... pmap already reports on process utilization.

**Comment by GuillaumeRochette** [comment]
I wanted to check because I maybe misunderstand how processes and threads are working.

To me, threads spawned from a process use the memory of the process they belong to.

So just to be sure, when I do the ```pmap -x``` of the main process, the grand total of the RSS column is equal to the amount of memory used by the main process and all of its spawned threads?

**Comment by jnwatson** [comment]
Yes, threads don't own any memory themselves. All the threads of a process share the same pool of memory. If one thread allocates memory, another thread in the same process can use that memory. For example, they all share the heap.

Yes, pmap -x shows memory usage of everything in the process, whoever allocated it.
